qrexec/tools/qrexec_legacy_convert.py

```diff
@@ -0,0 +1,309 @@
+            continue
+
+        missing = [rule for rule in legacy if str(rule) not in non_legacy_str]
+        if not missing:
+            continue
+
+        last_working_rule = len(missing)
+        for i, rule in enumerate(missing):
+            if str(rule.rule.action) == 'deny' \
+                    and str(rule.rule.source) == '@anyvm' \
+                    and str(rule.rule.target) == '@anyvm' \
+                    and not rule.rule.argument:
+                last_working_rule = i
+                break
+        missing = missing[:last_working_rule]
+        if not missing:
+            continue
+        filename = SERVICE_TO_FILE.get(service, CONFIG_FILE)
+
+        for rule in missing:
+            if rule.is_rule_compatible():
+                rules_to_save[filename].append(rule)
+            else:
+                rules_to_save[CONFIG_FILE].append(rule)
+
+    if '50-config-input' in rules_to_save:
+        config_file, user_file = \
+            split_input_rules(rules_to_save['50-config-input'])
+        rules_to_save['50-config-input'] = config_file
+        rules_to_save[CONFIG_FILE].extend(user_file)
+
+    backups_made: Dict[pathlib.Path, pathlib.Path] = {}
+
+    # do actual rule saving
+    for filename, rules in rules_to_save.items():
+        if not rules:
+            continue
+
+        file = POLICYPATH / (filename + ".policy")
+        disclaimer = DISCLAIMER if filename == '30-user' else TOOL_DISCLAIMER
+        text = disclaimer + '\n'.join([str(rule.rule) for rule in rules]) + '\n'
+
+        if file.exists():
+            backup_name = str(file) + '.bak'
+            while pathlib.Path(backup_name).exists():
+                backup_name = backup_name + '.bak'
+            backups_made[file] = pathlib.Path(backup_name)
+            shutil.copy(file, backup_name)
+
+            if filename != '50-config-input':
+                # input files are special: they replace current rules
+                current_text = file.read_text()
+                if current_text.startswith(TOOL_DISCLAIMER):
+                    current_text = current_text[len(TOOL_DISCLAIMER):]
+                text = text.rstrip('\n') + current_text
+
+        print("Writing " + str(file) + '...')
+        file.write_text(text)
+
+    # remove old
+    for file in POLICYPATH_OLD.iterdir():
+        if file.is_file() and not file.name.endswith('.rpmsave'):
+            backup_name = str(file) + '.rpmsave'
+            while pathlib.Path(backup_name).exists():
+                backup_name = backup_name + '.rpmsave'
+            backups_made[file] = pathlib.Path(backup_name)
+            file.rename(backup_name)
+
+    # check if state changed
+    current_state_string = StringIO()
+    qrexec_policy_graph.main(['--policy-dir',
+                              str(POLICYPATH), '--full-output'],
+                             output=current_state_string)
+    current_state = set(current_state_string.getvalue().split('\n'))
```

Comment:
Wrap this in some try-except (and in case of error set `current_state` to some value, like empty or "ERROR"). Otherwise, if the tool produce broken policy for some reason (see comment above), it won't offer to rollback.